### DISCUSSION TITLE
Enable End-to-End Payload Encryption

### DISCUSSION THREAD

**Discussion by anil-open** [filamentphp/filament#14628] [Help]
### Package

Panel builder

### Package Version

v3.2.121

### How can we help you?

Hello Filament Community!

I’d like to share and discuss an approach for adding end-to-end payload encryption on top of HTTPS for a Laravel 11 application using Filament v3. This setup adds an additional layer of security by encrypting specific client-server payloads, keeping sensitive data secure even within HTTPS-protected requests.

**Key Features:**
**Custom JavaScript for Client-Side Encryption:** A javascript module(class) that encrypts outgoing requests and decrypts incoming responses for specified endpoints using Web Crypto API.
**Middleware for Server-Side Decryption/Encryption:** Laravel middleware decrypts incoming payloads and encrypts outgoing responses, providing seamless integration with the backend.
**Configurable Encryption:** An environment variable toggles encryption on or off, making it adaptable to different environments or deployments.
**Key Exchange Mechanism:** On the first load, the application fetches a public encryption key from the server, used by the custom javascript module to handle encryption.
**Path Whitelisting:** Certain paths can bypass encryption to avoid unnecessary processing on endpoints that don’t require this extra layer of security.

The only thing I'm struggling with is how to tap into the Livewire requests and responses to change the outgoing payload to encrypted one and change the incoming response to decrypted one. Where should I wire up the frontend code? I have looked into the Livewire docs and have tried customizing the [Request Hook](https://livewire.laravel.com/docs/javascript#request-hooks)

`Relevant snippet of the custom javascript module:`

```javascript
// Initialize encryption service and register Livewire hooks
document.addEventListener('DOMContentLoaded', async () => {
    // Create a global instance of EncryptionService
    window.encryptionService = new EncryptionService();

    // Wait for Filament/Livewire to be ready
    await window.$wire?.initialRender;

    // Initialize encryption service
    await window.encryptionService.initialize();

    // Register Livewire request hook with encryption and decryption
    Livewire.hook('request', async ({ uri, options, payload, respond, succeed, fail }) => {
        // Encrypt request payload if encryption is enabled
        if (window.encryptionService.isEnabled && window.encryptionService.sessionId && payload) {
            try {
                // Set headers to identify the encrypted request
                options.headers = options.headers || {};
                options.headers['X-Session-Id'] = window.encryptionService.sessionId;
                options.headers['X-Encrypted'] = '1';

                // Encrypt the payload
                const encryptedPayload = await window.encryptionService.encryptPayload(payload);

                // Overwrite the payload with the encrypted content
                options.body = JSON.stringify({encrypted: encryptedPayload});

            } catch (error) {
                console.error('Encryption failed:', error);
                // If encryption fails, proceed without encryption
                payload = JSON.stringify(payload);
            }
        }

        // Handle the response
        respond(({ status, response }) => {
            // Check if the response is encrypted by looking at the header
            const isEncrypted = response.headers.get('X-Encrypted') === '1';

            // If encrypted, proceed to decrypt after parsing the raw response
            if (isEncrypted && window.encryptionService.sessionId) {
                response.text().then(async (encryptedContent) => {
                    try {
                        const decryptedContent = await window.encryptionService.decryptPayload(encryptedContent);
                        succeed({ status, json: decryptedContent });
                    } catch (error) {
                        console.error('Decryption failed:', error);
                        fail({ status, content: 'Decryption failed' });
                    }
                }).catch((error) => {
                    console.error('Error reading response content:', error);
                    fail({ status, content: 'Failed to read encrypted response content' });
                });
            } else {
                // If the response isn't encrypted, handle it as usual
                succeed({ status, json: response.json() });
            }
        });

        // If decryption fails or if there’s an error, handle it in `fail`
        fail(({ status, content, preventDefault }) => {
            console.error('Request failed:', status, content);
            if (window.$wireui) {
                window.$wireui.notification({
                    title: 'Error',
                    description: 'Failed to process request',
                    icon: 'error'
                });
            }
        });
    });
});
```

But this doesn't work and still sends the normal JSON payload to backend even though the above javascript is properly executed. Any pointers in right direction would be helpful.

Edit: Used `options.body` to modify the request body

**Comment by anil-open** [comment]
Hi @danharrin, is it possible for you take a look at the above and suggest on the approach?

**Reply by leandrocfe** [reply]
Does the code work if you test it in a standalone Livewire project?

**Reply by anil-open** [reply]
I figured out the issue. The Livewire request trigger is a non-async one as described [here](https://github.com/livewire/livewire/blob/main/js/request/index.js#L72) and I used it like below:

`Livewire.hook('request', async ({ uri, options, payload, respond, succeed, fail }) => {` 
because I needed to the call the encryption service async'ly. When I remove the `async` keyword from the callback function in the hook, it complains about the inner await but the payload actually gets modified. 

Is there a way to call an async function from within the hook? I tried wrapping the encryption function call in a promise but that didn't work.

**Reply by leandrocfe** [reply]
I don't know, but I think this isn't a Filament specifc issue. If it works with Livewire, it should also be compatible with Filament

**Reply by anil-open** [reply]
I'm not saying that there's an issue of any kind, the reason this thread is tagged under HELP is exactly because I need help/guidance to make this work. I've also raised the same thread in Livewire discussions but not getting any response there.
